src/db/users/users.service.ts

```diff
@@ -1,38 +1,59 @@
 import { Injectable, ConflictException, BadRequestException, InternalServerErrorException } from '@nestjs/common';
 import { InjectModel } from '@nestjs/mongoose';
 import { Model } from 'mongoose';
-import { isNullOrUndefined } from 'util';
 import { hash, argon2id } from 'argon2';
 import * as sanitize from 'sanitize-html';
 import validator from 'validator';
-
 import * as models from './models';
+import { isNullOrUndefined } from '../../util/validation';
 
 @Injectable()
 export class UsersService {
-    constructor(@InjectModel('User') private readonly userModel: Model<models.User>, @InjectModel('InviteCodes') private readonly inviteCodesModel: Model<models.InviteCodes>) {}
+    constructor(@InjectModel('User') private readonly userModel: Model<models.User>,
+        @InjectModel('InviteCodes') private readonly inviteCodesModel: Model<models.InviteCodes>) { }
 
     /**
-     * Creates a user and adds them to the database. First, it checks to see if
-     * a user with the requested username and/or email already exist. If true,
+     * Creates a user and adds them to the database. First, it ensures the 
+     * invite code is valid, and hasn't been used yet. Then, it checks to see if
+     * a user with the requested username and/or email already exists. If they do,
      * then the creation is rejected. Otherwise, the creation proceeds.
      * 
      * @param newUserInfo A new user's information
      */
     async createUser(newUserInfo: models.CreateUser): Promise<models.User> {
-        if (validator.isEmail(newUserInfo.email)) {
-            const existingUsername = await this.userModel.findOne({username: sanitize(newUserInfo.username)});
-            const existingEmail = await this.userModel.findOne({email: sanitize(newUserInfo.email)});
-
-            if (isNullOrUndefined(existingUsername) && isNullOrUndefined(existingEmail)) {
-                const newUser = new this.userModel(newUserInfo);
-                return await newUser.save();         
-            } else {
-                throw new ConflictException('Someone already has your username or email. Try another combination.');
-            }
-        } else {
+        if (!newUserInfo.inviteCode) {
```

Comment:
In addition to the extra validation, I reworked this to use the early return pattern. I tend to prefer it in validation-heavy stuff, because then the success value is always at the very bottom of the function.